_posts/2021-07-22-how-many-bootstrap-samples.md

```diff
@@ -0,0 +1,165 @@
+
+So, how many bootstrap samples do we actually need?
+
+## Finding the minimum number of samples
+
+To answer this question, we can keep adding more bootstrap samples until our uncertainty estimates stop changing significantly with each new sample. This is a one-time, up-front cost that will give us a good estimate of the minimum number of bootstrap samples required for our application. Once we have this, we can use that number each time we need to generate a bootstrapped uncertainty estimate, and avoid the cost & time of generating more samples than necessary at each model iteration.
+
+### Generating bootstrap samples
+
+To generate a bunch of samples of our ATT metric, we can do the following:
+
+1. Randomly sample our dataset (with replacement) to generate a new one
+2. Fit our causal model using this dataset
+3. Filter our dataset to the treated population
+4. Use our fitted model to predict the outcome when they receive the treatment
+5. Use our fitted model to predict the outcome when they DO NOT receive the treatment
+6. Calculate the difference for each record, and then average them to get ATT
+7. Write the ATT value to a file (so we don't have to redo this again later!)
+8. Repeat steps 1-7 10,000 times
+
+The code for 1-7 can be wrapped into a single function like this:
+
+```python
+def fit_model_and_get_att(filename, df, iteration):
+    # Randomly sample data
+    model_df = df.sample(replace=True, n=df.shape[0], random_state=iteration)
+
+    # Fit causal model
+    model_formula = 'is_purchase_made ~ confounder_1 + confounder_2 + treatment_variable'
+    _, X = dmatrices(
+                (model_formula), 
+                model_df, 
+                return_type = 'dataframe'
+            )
+    y = model_df.is_completed.astype(float)
+
+    model = LogisticRegression(random_state=0, max_iter=1000)
+    fitted_model = model.fit(X, y)
+
+    # Predict what would have happened WITH treatment
+    treated_df = X[X['treatment_variable[T.True]'] == 1].copy()
+    treated_preds = fitted_model.predict_proba(treated_df)[:, 1]
+
+    # Predict what would have happend WITHOUT treatment
+    treated_df_no_treatment = treated_df.copy()
+    treated_df_no_treatment['treatment_variable[T.True]'] = 0
+    treated_no_treatment_preds = fitted_model.predict_proba(treated_df_no_treatment)[:, 1]
+
+    # Calculate average treatment effect on the treated (att) & save to file!
+    treatment_effect_on_treated = treated_preds - treated_no_treatment_preds
+    att = treatment_effect_on_treated.mean()
+    with open(filename, "a") as f_out:
+        f_out.write(f"{att}\n")
+
+    return att
+```
+
+And then instead of running this function sequentially 10,000 times (slow!), we can leverage [Dask](https://dask.org/) to distribute this work across all available CPU cores (my Macbook Pro has 16).
```

Comment:
```suggestion
And then instead of running this function sequentially 10,000 times (slow!), we can leverage [Dask](https://dask.org/) to distribute this work across all available CPU cores (less slow!).
```